tower-http/src/auth/require_authorization_async.rs

```diff
@@ -0,0 +1,357 @@
+where
+    Auth: AuthorizeRequestAsync,
+    S: Service<Request<ReqBody>>,
+{
+    auth: Auth,
+    #[pin]
+    state: State<Auth::Future, ReqBody, S::Future>,
+    service: S,
+}
+
+impl<Auth, S, ReqBody, B> Future for ResponseFuture<Auth, S, ReqBody>
+where
+    Auth: AuthorizeRequestAsync<ResponseBody = B>,
+    S: Service<Request<ReqBody>, Response = Response<B>>,
+{
+    type Output = Result<Response<B>, S::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        loop {
+            match this.state.as_mut().project() {
+                StateProj::Authorize { authorize, req } => {
+                    let auth = ready!(authorize.poll(cx));
+                    let mut req = req.take().expect("future polled after completion");
+                    match auth {
+                        Some(output) => {
+                            this.auth.on_authorized(&mut req, output);
+                            let fut = this.service.call(req);
+                            this.state.set(State::Authorized { fut })
+                        }
+                        None => {
+                            let res = this.auth.unauthorized_response(&req);
+                            return Poll::Ready(Ok(res));
+                        }
+                    };
+                }
+                StateProj::Authorized { fut } => {
+                    return fut.poll(cx);
+                }
+            }
+        }
+    }
+}
+
+/// Trait for authorizing requests.
+pub trait AuthorizeRequestAsync {
+    /// The output type of doing the authorization.
+    ///
+    /// Use `()` if authorization doesn't produce any meaningful output.
+    type Output;
+
+    /// The Future type returned by `authorize`
+    type Future: Future<Output = Option<Self::Output>>;
+
+    /// The body type used for responses to unauthorized requests.
+    type ResponseBody: Body;
+
+    /// Authorize the request.
+    ///
+    /// If `Some(_)` is returned then the request is allowed through, otherwise not.
+    fn authorize<B>(&mut self, request: &Request<B>) -> Self::Future;
+
+    /// Callback for when a request has been successfully authorized.
+    ///
+    /// For example this allows you to save `Self::Output` in a [request extension][] to make it
+    /// available to services further down the stack. This could for example be the "claims" for a
+    /// valid [JWT].
+    ///
+    /// Defaults to doing nothing.
+    ///
+    /// See the [module docs](crate::auth::require_authorization_async) for an example.
+    ///
+    /// [request extension]: https://docs.rs/http/latest/http/struct.Extensions.html
+    /// [JWT]: https://jwt.io
+    #[inline]
+    fn on_authorized<B>(&mut self, _request: &mut Request<B>, _output: Self::Output) {}
+
+    /// Create the response for an unauthorized request.
+    fn unauthorized_response<B>(&mut self, request: &Request<B>) -> Response<Self::ResponseBody>;
```

Comment:
Should there be a way to pass the error from authorize function to this? The use case for that is, if there is a db or redis call in authorize function, then on network error we should be able to send a 5xx error instead of a 401 error.

Reply:
I agree with you but IMO that's outside the scope of this PR. This PR adds async version of `RequireAuthorization`. Such a change would also need to made to `RequireAuthorization` in addition to `RequireAuthorizationAsync`.

Reply:
I think mapping errors to responses should be handled by a separate middleware. So I think this is fine. We can always consider adding error handling to these middleware for 0.2, if someone requests it.

Reply:
...and that person will be me. I was using it and ran into a case where it's impossible to return 401/403 depending upon the request. We need a better way here.

I was gonna make an issue but life got in the way...

Reply:
Hehe alright. Do you wanna address it here or should we move forward with this as is, and fix it for 0.2? I'm fine with either actually.

Reply:
Let's fix it in some other PR. It's a breaking change. I don't have the time for it right now anyway.